### PR TITLE
Replace sparse __getattr__ with properties

### PR DESCRIPTION
**Pull request by ivirshup**
This PR replaces use of `__getattr__` in `_spbase` with properties. This should make code a bit more predictable, and generally "more pythonic/ standard"

### CODE REVIEW

File: scipy/sparse/_base.py [hunk 1]
```diff
@@ -711,31 +711,39 @@ def __itruediv__(self, other):
...
-        elif attr == 'size':
-            return self._getnnz()
-        else:
-            raise AttributeError(attr + " not found")
+    @property
+    def A(self) -> np.ndarray:
+        if self._is_array:
+            warn(np.VisibleDeprecationWarning(
+                "`.A` is deprecated and will be removed in v1.13.0. "
+                "Use `.todense()` instead."
```
Comment (rossbar) [inline scipy/sparse/_base.py:719]:
```suggestion
                "Use `.toarray()` instead."
```

Comment (ivirshup) [reply scipy/sparse/_base.py]: Changed!

**Comment by ivirshup** [comment]
Test failing due to:

```pytb
scipy/optimize/_linprog_ip.py:727: in _ip_hsd
    A.T = A.transpose()  # A.T is defined for sparse matrices but is slow
E   AttributeError: can't set attribute 'T'
```

Probably fixable cause:

```python
    @property
    def T(self):
        return self.transpose()
```

**Comment by ivirshup** [comment]
The offending line was in the "interior point" linear programming implementation and was trying to cache the transpose of the matrix, saying that it's slow. 

* The transpose of a csc matrix is currently quite fast (e.g., it's the same data as a csr matrix), so I think this could be outdated.
* This method is deprecated, and the benchmark has been removed

Word is that this makes it fine to just no longer cache.

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
This LGTM. In addition to making the code a bit cleaner, this also resolves some potential headaches due to the way Python swallows exceptions when a property throws an error but `__getattr__` is also defined. (See https://medium.com/@ceshine/python-debugging-pitfall-mixed-use-of-property-and-getattr-f89e0ede13f1 for example).

I'll hold off on merging until we get a +1 from someone more familiar with scipy.optimize (@andyfaff, @mdhaber maybe?)

**Comment by perimosocordiae** [comment]
We got a verbal ok from @mdhaber, so I'll merge this now.

**Pull request merged**